// @formatjs/cli-lib/src/formatters/smartling.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.compile = exports.compareMessages = exports.format = void 0;
const format = msgs => {
    const results = {
        smartling: {
            translate_paths: [
                {
                    path: '*/message',
                    key: '{*}/message',
                    instruction: '*/description',
                },
            ],
            variants_enabled: true,
            string_format: 'icu',
        },
    };
    for (const [id, msg] of Object.entries(msgs)) {
        results[id] = {
            message: msg.defaultMessage,
            description: typeof msg.description === 'string'
                ? msg.description
                : JSON.stringify(msg.description),
        };
    }
    return results;
};
exports.format = format;
const compareMessages = (el1, el2) => {
    // `smartling` has to be the 1st key
    if (el1.key === 'smartling') {
        return -1;
    }
    if (el2.key === 'smartling') {
        return 1;
    }
    return el1.key < el2.key ? -1 : el1.key === el2.key ? 0 : 1;
};
exports.compareMessages = compareMessages;
const compile = msgs => {
    const results = {};
    for (const [id, msg] of Object.entries(msgs)) {
        if (id === 'smartling') {
            continue;
        }
        results[id] = msg.message;
    }
    return results;
};
exports.compile = compile;
